File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-10.java
class Solution {
    Integer[][] dp;
    int[][] dir = {{0,1},{1,0},{0,-1},{-1,0}};
    int M,N;
    public int longestIncreasingPath(int[][] matrix) {
        M = matrix.length;
        N = matrix[0].length;
        dp = new Integer[M][N];
        int res = 0;
        for(int i =0;i<M;i++){
            for(int j = 0;j<N;j++){
                res = Math.max(res,getMax(i,j,matrix));
            }
        }
        return res;
        
    }

    public int getMax(int i,int j,int[][] matrix){
        int res = 1;
        if(dp[i][j]!=null){
            return dp[i][j];
        }
        int nei = 0;
        for(int[] d:dir){
            int x = d[0]+i;
            int y = d[1]+j;
            if(x<0 || x==M ||y<0 || y==N || matrix[x][y]<=matrix[i][j]){
                continue;
            }
            nei= Math.max(nei,getMax(x,y,matrix));
        }
        dp[i][j] = res+nei;
        return dp[i][j];
    }
}

File: Data Structures & Algorithms/jump-game-ii/submission-10.py
class Solution:
    def jump(self, nums: List[int]) -> int:

        res = 0
        l,r = 0,0
        while r<len(nums)-1:
            res+=1
            nextR = r
            for i in range(l,r+1):
                nextR = max(nextR,i+nums[i])
            l = r+1
            r = nextR
        return res


            


        # 2   4   1   1   1   1

        # [2] 1 

        # [4,1] 2 

        